File: src/main/java/com/example/chatService/service/ChatMessageService.java
package com.example.chatService.service;


import com.example.chatService.component.ChatRateLimiter;
import com.example.chatService.dto.GroupMessageDto;
import com.example.chatService.kafka.GroupMessageProducer;
import lombok.RequiredArgsConstructor;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
@RequiredArgsConstructor
public class ChatMessageService {

    private final ChatRateLimiter chatRateLimiter;
    private final SimpMessagingTemplate messagingTemplate;
    private final GroupMessageProducer groupMessageProducer;

    public void send(GroupMessageDto msg, String senderIdFromPrincipal) {

        msg.setSenderId(senderIdFromPrincipal);

        if (!chatRateLimiter.allowUser(senderIdFromPrincipal)) {
            return; // drop
        }

        if (!chatRateLimiter.allowRoom(msg.getRoomId())) {
            return; // drop
        }

        // 밴이면 알림 보내고 차단
        if (!chatRateLimiter.allowOrBan(senderIdFromPrincipal)) {
            long ttl = chatRateLimiter.getBanTtl(senderIdFromPrincipal);

            messagingTemplate.convertAndSendToUser(
                    senderIdFromPrincipal,
                    "/queue/rate-limit",
                    Map.of(
                            "type", "CHAT_BANNED",
                            "retryAfter", ttl
                    )
            );
            return;
        }
        groupMessageProducer.send(msg);
    }
}

File: src/main/java/com/example/chatService/service/ChatMessageService.kt
package com.example.chatService.service

import com.example.chatService.component.ChatRateLimiter
import com.example.chatService.dto.GroupMessageDto
import com.example.chatService.kafka.GroupMessageProducer
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Service

@Service
class ChatMessageService(
        private val chatRateLimiter: ChatRateLimiter,
        private val messagingTemplate : SimpMessagingTemplate,
        private val groupMessageProducer: GroupMessageProducer
) {
    fun send(msg : GroupMessageDto, senderIdFromPrincipal : String) {
        msg.senderId = senderIdFromPrincipal

        if (!chatRateLimiter.allowUser(senderIdFromPrincipal)) {
            return
        }
        if (!chatRateLimiter.allowRoom(msg.roomId)) {
            return
        }

        // 밴이면 알림 보내고 차단
        if (!chatRateLimiter.allowOrBan(senderIdFromPrincipal)) {
            val ttl = chatRateLimiter.getBanTtl(senderIdFromPrincipal)
            messagingTemplate.convertAndSendToUser(
                    senderIdFromPrincipal,
                    "/queue/rate-limit",
                    mapOf(
                            "type" to "CHAT_BANNED",
                            "retryAfter" to ttl
                    )
            )
            return
        }
        groupMessageProducer.send(msg)
    }
}